lib/misc.js

```diff
@@ -63,5 +65,91 @@ exports.generateId = function(fields) {
   hasher.update('_');
 
   return base32hex.encodeBuffer(hasher.digest(), { paddingChar: '-' });
-};
+}
+
+function generateInternalId(externalId, groupId) {
+  return generateId([externalId, groupId]);
+}
+
+function ensureInternalId(datum) {
+  if (!datum.id) {
+    datum.id = generateExternalId(datum);
+  }
+  if (!datum._id) {
+    datum._id = generateInternalId(datum.id, datum._groupId);
+  }
+  return datum;
+}
+
+function generateExternalId(datum) {
+  return schema.generateId(datum, schema.idFields(datum.type));
+}
+
+function sortTimeAsc(a, b) {
+  const timeA = new Date(a.time);
+  const timeB = new Date(b.time);
+  if (timeA < timeB)
+    return -1;
+  if (timeA > timeB)
+    return 1;
+  return 0;
+}
+
+/**
+ * @typedef {Object} RemoveDuplicatesResult
+ * @property {Object[]} array - the non duplicates that should be persisted to the database.
+ * @property {Object[]} duplicates - duplicate array entries that should not be persisted (or would return an error if they would).
+ * @property {Object} latestData - a convenience mapping of [_id] => datum
+ */
+
+/**
+ * Returns data that should be inserted into the database taking into account existing data.
+ *
+ * @param {Object[]} existingDBData - Array of data already existing in the database.
+ * @param {Object[]} inputData - Array of data that user intends to save to the database.
+ * @return {RemoveDuplicatesResult}
+ */
+function removeDuplicates(existingDBData, inputData) {
+  const latestData = {};
+  for (const datum of existingDBData) {
+    const externalId = generateExternalId(datum);
+    const internalId = generateInternalId(externalId, datum._groupId);
+    latestData[datum._id] = datum;
+  }
+
+  // Duplicate inserted data may exist in the inputData array. This occurs
+  // when there is multiple data with the same id / _id. This is not allowed
+  // and in the serial code would return a { statusCode: 400,
+  // errorCode: 'duplicate', message: 'received a duplicate event'} error to
+  // the callback. Since there may be multiple duplicates we instead include
+  // an array of duplicates in the returned object.
+  const finalArray = [];
+  const duplicates = [];
+  inputData = inputData.slice(0);
+  inputData.sort(sortTimeAsc);
+  for (const datum of inputData) {
+    const externalId = generateExternalId(datum);
+    const internalId = generateInternalId(externalId, datum._groupId);
+
+    // The absence of datum.createdTime indicates an insert whereas the
+    // presence of it indicates an update. Disallow multiple inserts.
+    if (!datum.createdTime && latestData[internalId]) {
+      duplicates.push(datum);
+      continue;
+    }
+    latestData[internalId] = datum;
+    finalArray.push(datum);
```

Comment:
I believe this will create duplicate data in the database under certain, though not normal, circumstances. For example, if the input data has two updates (where `datum.createTime` is specified) for the same internal id.

I don't think this will happen under normal circumstances with the uploader, but who knows with third parties.

Reply:
Any datums with the same internal id will fail on the insert because the primary key _id will be the same so I think we're ok here - the old code relied on this behaviour as well.